runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -512,10 +512,24 @@ void J9::RecognizedCallTransformer::process_java_lang_StrictMath_and_Math_sqrt(T
 /*
 Transform an Unsafe atomic call to diamonds with equivalent semantics
 
+If OffHeap is enabled, a runtime isArray check is added to determine
+if loading the dataAddrPtr is necessary.
+
                           yes
 isObjectNull [A] ------------------------------------------>
     |                                                      |
     | no                                                   |
+#if OffHeap                                                |
+    |                yes                                   |
+isArray [I] ------------------------>                      |
```

Comment:
I supposed array-class(es) don't have associated static fields. otherwise, it seems possible that isArray & isLowTagged are true at the same time? then, this scheme might not work.  i.e.  this scheme assumes: when isArray is true, isLowTagged cannot be true. the assumption is reasonable to me, you had better double-check to be sure with @tajila

Reply:
I added the comments for some of the code that @midronij did, so just to make sure that I got it right.
@midronij can you confirm that the comments fit the code you made regarding this?

Reply:
The comments look good! 

In general, it seems like the way this is handled is very similar to `Unsafe.get()` and `Unsafe.put()` (performing a NULL test, array test, and lowtag test at runtime to determine how the dst/src addresses should be calculated), so unless there is some key difference between those and `Unsafe.getAndAdd()`, I think it should be safe to assume that isLowTagged and isArray can't be true at the same time. But as Julian said, Tobi is probably the best person to answer that.